problem_builder/instructor_tool.py

```diff
@@ -0,0 +1,359 @@
+                    "id": block_id,
+                    "name": block_name,
+                    "eligible": eligible,
+                }
+                flat_block_tree.append(new_entry)
+                if block.has_children and not getattr(block, "has_dynamic_children", lambda: False)():
+                    for child_id in block.children:
+                        build_tree(block.runtime.get_block(child_id), ancestors=(ancestors + [new_entry]))
+
+        root_block = self
+        while root_block.parent:
+            root_block = root_block.get_parent()
+        root_block_id = get_block_id(root_block)
+        root_entry = {
+            "depth": 0,
+            "id": root_block_id,
+            "name": "All",
+            "eligible": False,
+        }
+        flat_block_tree.append(root_entry)
+
+        for child_id in root_block.children:
+            child_block = root_block.runtime.get_block(child_id)
+            build_tree(child_block, [root_entry])
+
+        return flat_block_tree
+
+    @property
+    def download_url_for_last_report(self):
+        """ Get the URL for the last report, if any """
+        # Unfortunately this is a bit inefficient due to the ReportStore API
+        if not self.last_export_result or self.last_export_result['error'] is not None:
+            return None
+        from instructor_task.models import ReportStore
+        report_store = ReportStore.from_config(config_name='GRADES_DOWNLOAD')
+        course_key = getattr(self.scope_ids.usage_id, 'course_key', None)
+        return dict(report_store.links_for(course_key)).get(self.last_export_result['report_filename'])
+
+    def _get_status(self):
+        self.check_pending_export()
+        return {
+            'export_pending': bool(self.active_export_task_id),
+            'last_export_result': self.last_export_result,
+            'download_url': self.download_url_for_last_report,
+        }
+
+    def raise_error(self, code, message):
+        """
+        Raises an error and marks the block with a simulated failed task dict.
+        """
+        self.last_export_result = {
+            'error': message,
+        }
+        self.display_data = None
+        raise JsonHandlerError(code, message)
+
+    @XBlock.json_handler
+    def get_status(self, data, suffix=''):
+        return self._get_status()
+
+    @XBlock.json_handler
+    def delete_export(self, data, suffix=''):
+        self._delete_export()
+        return self._get_status()
+
+    def _delete_export(self):
+        self.last_export_result = None
+        self.display_data = None
+        self.active_export_task_id = ''
+
+    @XBlock.json_handler
+    def start_export(self, data, suffix=''):
+        """ Start a new asynchronous export """
+        block_types = data.get('block_types', None)
+        usernames = data.get('usernames', None)
+        root_block_id = data.get('root_block_id', None)
+        match_string = data.get('match_string', None)
```

Comment:
@dianakhuang CC @ormsbee

That's a valid concern and good to be aware of.  I believe this code is safe, but please let me know if my assessment is incorrect?

The blocks returned are fetched via [`modulestore.get_items()`](https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/modulestore/mixed.py#L261-L291), using the modulestore identified by the current runtime's `course_id`.  See the call to [`tasks.export_data()`](https://github.com/open-craft/problem-builder/blob/432c1b9a7a42b0d405de83acacee6b1ad4879b69/problem_builder/instructor_tool.py#L320-L327), and the [`export_data` task](https://github.com/open-craft/problem-builder/blob/432c1b9a7a42b0d405de83acacee6b1ad4879b69/problem_builder/tasks.py#L22-L32) itself.